trio/_threads.py

```diff
@@ -240,37 +315,64 @@ def deliver_worker_fn_result(result: outcome.Outcome[RetT]) -> None:
         limiter.release_on_behalf_of(placeholder)
         raise
 
-    def abort(_: RaiseCancelT) -> trio.lowlevel.Abort:
+    def abort(raise_cancel: RaiseCancelT) -> trio.lowlevel.Abort:
+        # fill so from_thread_check_cancelled can raise
+        cancel_register[0] = raise_cancel
         if cancellable:
+            # empty so report_back_in_trio_thread_fn cannot reschedule
             task_register[0] = None
             return trio.lowlevel.Abort.SUCCEEDED
         else:
             return trio.lowlevel.Abort.FAILED
 
-    # wait_task_rescheduled return value cannot be typed
-    return await trio.lowlevel.wait_task_rescheduled(abort)  # type: ignore[no-any-return]
+    while True:
+        # wait_task_rescheduled return value cannot be typed
+        msg_from_thread: ThreadDone[RetT] | Run[RetT] | RunSync[
+            RetT
+        ] = await trio.lowlevel.wait_task_rescheduled(abort)
+        if type(msg_from_thread) is ThreadDone:
+            return msg_from_thread.result.unwrap()  # type: ignore[no-any-return]
+        elif type(msg_from_thread) is Run:
+            await msg_from_thread.run()
+        elif type(msg_from_thread) is RunSync:
+            msg_from_thread.run_sync()
+        else:  # pragma: no cover, internal debugging guard
+            raise TypeError(
```

Comment:
I like this, but are we doing `assert_never` anywhere else in the library? otherwise I think we could hold off here and do it in multiple places as a separate PR.

Reply:
We can’t be no, since we’d need to copy it in as I mentioned. We can’t directly use the `typing_extensions` back port since we’re avoiding a runtime dependency.

Reply:
> If you add `@typing.final` to all three classes (disallowing subclassing), we’d then be able to use `typing.assert_never()` here to have the checker prove it’s impossible. We can just copy the definition for earlier Python versions into utils:

Side note, i'm using `type(x) is Type` rather than `isinstance`, doesn't that also disallow subclasses?

Reply:
Not quite, because if someone did pass a subclass of one of the types in, it’d fail all the if checks since it isn’t one of those specific types.